pkg/openfeature/event_executor.go

```diff
@@ -0,0 +1,319 @@
+package openfeature
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"golang.org/x/sync/errgroup"
+)
+
+// event executor is a registry to connect API and Client event handlers to Providers
+
+// handlerExecutionTime defines the maximum time event handler will wait for its handlers to complete
+const handlerExecutionTime = 500 * time.Millisecond
+
+type eventExecutor struct {
+	defaultProviderReference *providerReference
+	namedProviderReference   map[string]*providerReference
+	apiRegistry              map[EventType][]EventCallback
+	scopedRegistry           map[string]scopedCallback
+	logger                   logr.Logger
+	mu                       sync.Mutex
+}
+
+func newEventExecutor(logger logr.Logger) eventExecutor {
+	return eventExecutor{
+		namedProviderReference: map[string]*providerReference{},
+		apiRegistry:            map[EventType][]EventCallback{},
+		scopedRegistry:         map[string]scopedCallback{},
+		logger:                 logger,
+	}
+}
+
+// scopedCallback is a helper struct to hold client name associated callbacks.
+// Here, the scope correlates to the client and provider name
+type scopedCallback struct {
+	scope     string
+	callbacks map[EventType][]EventCallback
+}
+
+func newScopedCallback(client string) scopedCallback {
+	return scopedCallback{
+		scope:     client,
+		callbacks: map[EventType][]EventCallback{},
+	}
+}
+
+type providerReference struct {
+	eventHandler          *EventHandler
+	metadata              Metadata
+	clientNameAssociation string
+	isDefault             bool
+	shutdownSemaphore     chan interface{}
+}
+
+// updateLogger updates the executor's logger
+func (e *eventExecutor) updateLogger(l logr.Logger) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	e.logger = l
+}
+
+// registerApiHandler add API(global) level handler
```

Comment:
```suggestion
// registerApiHandler adds an API(global) level handler
```

I've seen this in a couple places.

I don't know if there's a convention around this, but if you keep referencing the name in the godoc, you'll have to conjugate all these verbs with an "s". The other option is just to drop the method name in the godoc, I guess:

`registerApiHandler adds an API(global) level handler`

vs

`add an API(global) level handler`

Not sure which is better.

Reply:
Referring the function/struct name as the first word (especially for exposed ones) is a convention in go [1]

I will revisit the comments and fix the grammar 

[1] - https://go.dev/doc/comment

Reply:
Updated with the latest commit